source/server/config/network/http_connection_manager.cc

```diff
@@ -67,6 +67,22 @@ NetworkFilterFactoryCb HttpConnectionManagerFilterConfigFactory::createFilterFac
   };
 }
 
+} // namespace
+
+NetworkFilterFactoryCb HttpConnectionManagerFilterConfigFactory::createFilterFactory(
+    const Json::Object& json_http_connection_manager, FactoryContext& context) {
+  envoy::api::v2::filter::HttpConnectionManager http_connection_manager;
+  Config::FilterJson::translateHttpConnectionManager(json_http_connection_manager,
+                                                     http_connection_manager);
+  return createHttpConnectionManagerFilterFactory(http_connection_manager, context);
+}
+
+NetworkFilterFactoryCb HttpConnectionManagerFilterConfigFactory::createFilterFactoryFromProto(
+    const Protobuf::Message& config, FactoryContext& context) {
+  return createHttpConnectionManagerFilterFactory(
+      dynamic_cast<const envoy::api::v2::filter::HttpConnectionManager&>(config), context);
```

Comment:
Just to check my understanding here: For a struct, whether JSON or proto, the type if embedded, so when it is actually parsed, it becomes the real object? Is that right?

Reply:
Does this mean that HttpConnectionManager message actually never appears on the wire, but is used internally by Envoy only? I.e., to pass in http connection manager config via a protobuf I have to map from the HttpConnectionManager message definition to a protobuf Struct, field by field. If so, what is the point of specifying the HttpConnectionMananger message in the envoy-api?

Reply:
Could not figure it out, managed to get it to work by encoding to Struct, seems to work but looks quite inefficient coding-wise:
```
    filter_chains {
      filters {
        name: "http_connection_manager"
        config {
          fields {
            key: "rds"
            value {
              struct_value {
                fields {
                  key: "config_source"
                  value {
                    struct_value {
                      fields {
                        key: "api_config_source"
                        value {
                          struct_value {
                            fields {
                              key: "api_type"
                              value {
                                number_value: 2
                              }
                            }
                            fields {
                              key: "cluster_name"
                              value {
                                string_value: "rdsCluster"
                              }
                            }
                          }
                        }
                      }
                    }
                  }
                }
                fields {
                  key: "route_config_name"
                  value {
                    string_value: "router1"
                  }
                }
              }
            }
          }
          fields {
            key: "stat_prefix"
            value {
              string_value: "proxy"
            }
          }
        }
        deprecated_v1 {
          type: "read"
        }
      }
    }
```

Reply:
Yes, it only appears as a `Struct`. What you have is correct. The goal here is not efficiency, but being extensible, we don't want to predefine all filter config protos upfront, we want to support user extension. The alternative `google.protobuf.Any` type was not attractive as it is hard to do anything with the encoded proto without the original schema, e.g. dumping debug, assembling in config generation tools.

This is why we use JSON/YAML for textual representation, the ugliness of `Struct` is hidden.

Reply:
We could have both by having the generic Struct as one of the Any alternatives, the others being protos specified as part of the envoy api. Extensibility would be as good as today, but dealing with "upstreamed" filters would be much more efficient. Or else remove the http_connection_manager.proto from envoy-api, as it clearly is not part of the gRPC API. Or are you saying that http_connection_manager.proto can be automatically converted to a JSON/YAML schema and those can be used as-is for v2 REST or REST_LEGACY ApiTypes?

Reply:
I'm interested in understanding the efficiency requirements here. Two questions:
1. Given this is control plane, I wouldn't normally be that concerned about slight encoding inefficiency. Do you have some very significant scaling requirements in terms of the size/frequency of config update?
2. I haven't looked at the wire format for Struct, but I assume the main objection is the fact that the key string is included, which is not needed if you know the proto upfront.
TBH, this is part of the API which we've tentatively frozen, so it would take a compelling reason to make this change.

Reply:
I don't have any numbers now, but you are likely right in that performance difference does not matter here. Even so it seems to me that there is much bigger difference in encoding/decoding CPU/memory use than in the wire format. Given that one of the main rationales of protobufs vs. XML (for example) is >10x encoding/decoding performance difference this seems like going backwards.

On the second point the change could be made backwards compatible by keeping the code point 2 for Struct config, and adding new code point(s) for Any or Oneof Message configs?

Reply:
I think the best thing to do at this point is open an issue and let's see what others in the community think.

Reply:
@PiotrSikora

Reply:
See https://github.com/envoyproxy/envoy/issues/1680.